statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,334 @@
+}
+
+// FreshMemUsage re-calculate the memory message
+func (l *internalLRUCache) FreshMemUsage() {
+	var trackingCost int64
+	var totalCost int64
+	l.Lock()
+	for _, v := range l.elements {
+		item := v.Value.(*cacheItem)
+		oldMemUsage := item.tblMemUsage
+		newMemUsage := item.value.MemoryUsage()
+		item.tblMemUsage = newMemUsage
+		l.calculateCost(newMemUsage, oldMemUsage)
+		l.maintainList(v, nil, newMemUsage, oldMemUsage)
+	}
+	l.evictIfNeeded()
+	totalCost = l.totalCost
+	trackingCost = l.trackingCost
+	l.Unlock()
+	metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+	metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+}
+
+// FreshTableCost re-calculate the memory message for the certain key
+func (l *internalLRUCache) FreshTableCost(key int64) {
+	var trackingCost int64
+	var totalCost int64
+	var calculated bool
+	l.Lock()
+	calculated = l.calculateTableCost(key)
+	totalCost = l.totalCost
+	trackingCost = l.trackingCost
+	l.Unlock()
+	if calculated {
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+	}
+}
+
+func (l *internalLRUCache) calculateTableCost(key int64) bool {
+	element, exists := l.elements[key]
+	if !exists {
+		return false
+	}
+	item := element.Value.(*cacheItem)
+	l.put(item.key, item.value, item.value.MemoryUsage(), true)
+	return true
+}
+
+// Copy returns a replication of LRU
+func (l *internalLRUCache) Copy() statsCacheInner {
+	var newCache *internalLRUCache
+	l.RLock()
+	newCache = newInternalLRUCache(l.capacity)
+	node := l.cache.Back()
+	for node != nil {
+		key := node.Value.(*cacheItem).key
+		value := node.Value.(*cacheItem).value
+		tblMemUsage := node.Value.(*cacheItem).tblMemUsage
+		newCache.put(key, value, tblMemUsage, false)
+		node = node.Prev()
+	}
+	l.RUnlock()
+	metrics.StatsCacheLRUCounter.WithLabelValues(typCopy).Inc()
+	return newCache
+}
+
+// internalLRUCache will evict a table's column' structure in order to keep tracking cost under capacity
+// If the elements has no structure can be evicted, it will be removed from list.
+func (l *internalLRUCache) evictIfNeeded() {
+	curr := l.cache.Back()
+	evicted := false
+	for l.trackingCost > l.capacity && curr != nil {
+		evicted = true
+		item := curr.Value.(*cacheItem)
+		tbl := item.value
+		oldMemUsage := item.tblMemUsage
+		prev := curr.Prev()
+		for _, col := range tbl.Columns {
+			col.DropEvicted()
+			newMemUsage := tbl.MemoryUsage()
+			item.tblMemUsage = newMemUsage
+			l.calculateCost(newMemUsage, oldMemUsage)
+			if l.trackingCost <= l.capacity {
+				break
+			}
+		}
+		newMemUsage := tbl.MemoryUsage()
+		if newMemUsage.TotalColTrackingMemUsage() < 1 {
+			l.maintainList(curr, nil, newMemUsage, oldMemUsage)
+		}
+		if l.trackingCost <= l.capacity {
+			break
+		}
+		curr = prev
+	}
+	if evicted {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typEvict).Inc()
```

Comment:
Do we need to record how many tables are evicted?

Reply:
Recording how many tables got evicted is necessary, but I think it's better to provide a api in statsCache instead of using metrics.